### PR TITLE
Restrict interleaving Hexagon pattern matches

### PR DESCRIPTION
**Pull request by shubhamp-ca**
This prevents matching interleaving hexagon patterns if the result can't be interleaved. Checks that the
op type is atleast 2 vector widths for widening ops.

Currently, compile-time assert is thrown in Codegen (see: https://github.com/halide/Halide/issues/1582) when the interleaved hexagon intrinsic op is not equal to two vector width. This is a recurring issue in widening ops, where the widening ops have to be vectorized by 2 * native vector width for valid codegen. This change lets the backend ISEL handle half-native vectors, instead of applying interleaved widening Hexagon intrinsics which the backend miscompiles.
@dsharletg PTAL

### CODE REVIEW

File: src/HexagonOptimize.cpp [hunk 1]
```diff
@@ -246,6 +246,11 @@ Expr replace_pattern(Expr x, const vector<Expr> &matches, const Pattern &p) {
     return x;
 }
 
+bool is_double_vector(Expr x, const Target &target) {
+    int native_vector_bits = target.natural_vector_size(Int(8)) * 8;
+    return (x.type().lanes() / ( native_vector_bits / x.type().bits()) >= 2);
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: I think this needs to be % 2 == 0, not >= 2? Or more clearly (IMO):

    int native_vector_lanes = target.natural_vector_size(x.type());
    return x.type.lanes() % (2 * native_vector_lanes) == 0;

Comment (shubhamp-ca) [reply src/HexagonOptimize.cpp]: Changed. Thanks.

**Comment by shubhamp-ca** [comment]
> I'm not seeing any false positives now, thanks!

If this is the case, I guess having is_double_vector() check before expr_match() is the better thing to do, as it saves the effort of having to pattern match an expression which shouldn't be. Which is to say -- `(expr_match && is_double_vector)`   is same as -- `(is_double_vector && expr_match)`, but the latter may be a bit more efficient?

File: src/HexagonOptimize.cpp [hunk 2]
```diff
@@ -269,6 +274,11 @@ Expr apply_patterns(Expr x, const vector<Pattern> &patterns, const Target &targe
                 continue;
             }
 
+            // Don't apply pattern if it does not involve an
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: This comment is a bit confusing. I think it should be rewritten "Don't apply pattern if it involves an interleave, and it is not a multiple of two vectors" and link to the issue #1582.

Comment (shubhamp-ca) [reply src/HexagonOptimize.cpp]: Done. Thanks.

**Comment by steven-johnson** [comment]
clone-for-build looks clean. ready to land?

**Pull request merged**